106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,507 @@
+With this proposal, we are only going to support auto-rebalance on imbalanced cluster.
+We also plan to implement the same for topic and metrics related issues, but it will be part of future work since their implementation require different approach.
+For example, when dealing with topic related issues, it will require a coordination with topic operator and metrics issues will require coordination with the Kafka API.
+
+Here is an example of what the configured `Kafka` custom resource could look like:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+  # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using the imbalance rebalance mode
+      - mode: imbalance
+        template:
+          name: my-imbalance-rebalance-template
+```
+
+It is also possible to use the default Cruise Control rebalancing configuration by omitting the `template` field.
+
+The configuration for Cruise Control rebalance is provided through a `KafkaRebalance` custom resource which is specifically defined as a template and is referenced in the `template` field in the example above.
+A `KafkaRebalance` custom resource with the `strimzi.io/rebalance-template: true` annotation set can be used as template.
+When a `KafkaRebalance` with this annotation is created, the cluster operator doesn't run any rebalancing.
+This is not an actual rebalance request to get an optimization proposal; it is simply where the configuration for auto-rebalancing is defined.
+
+Here is an example template:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-imbalance-rebalance-template
+  annotations:
+    strimzi.io/rebalance-template: "true" # specifies that this KafkaRebalance is a rebalance configuration template
+spec:
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+When the `template` is set, the operator automatically creates (or updates) a corresponding `KafkaRebalance` custom resource (based on the template) when an anomaly is detected and notified by the `StrimziCruiseControlNotifier`
+The operator copies over goals and rebalancing options from the referenced `template` resource to the generated rebalancing one.
+If the user has not configured the anomaly detection goals in Cruise Control then the operator would set the default goals to be used by the anomaly detector.
+The goals mentioned in the rebalance template will be validated with the anomaly detection goals being used by Cruise Control. The mentioned goals should be either same or be a subset of the anomaly detection goals since the unmentioned goals will not be detected and therefore no rebalance would be required for those goals.
```

Comment:
Wording suggestions to make this clearer. I'm making some assumptions here, so please validate that I've understood correctly.
```suggestion
If the user specifies a rebalance template then the goals mentioned in the rebalance template will be validated against the anomaly detection goals being used by Cruise Control.
The goals in the rebalance template should be either the same or be a subset of the anomaly detection goals being used by Cruise Control.
This is to ensure rebalances are only being run with goals that Cruise Control is checking for anomalies.
```

Reply:
Yes, you understood correct